addons/account_lock/models/res_company.py

```diff
@@ -1,13 +1,90 @@
 # -*- coding: utf-8 -*-
 
-from odoo import models, api
+import calendar
+from dateutil.relativedelta import relativedelta
+
+from odoo import fields, models, api, _
+from odoo.exceptions import ValidationError
 
 
 class ResCompany(models.Model):
     _inherit = 'res.company'
 
+    def _autorise_lock_date_changes(self, vals):
+        '''Check the lock dates for the current companies. This can't be done in a api.constrains because we need
+        to perform some comparison between new/old values. This method forces the lock dates to be irreversible.
+        * You cannot define stricter conditions on advisors than on users. Then, the lock date on advisor must be set
+        after the lock date for users.
+        * You cannot lock a period that is not finished yet. Then, the lock date for advisors must be set after the
+        last day of the previous month.
+        * The new lock date for advisors must be set after the previous lock date.
+        * You cannot delete a tax lock date, lock a period that is not finished yet or the tax lock date must be set after
+        the last day of the previous month.
+        :param vals: The values passed to the write method.
+        '''
+        period_lock_date = vals.get('period_lock_date') and\
+            fields.Date.from_string(vals['period_lock_date'])
+        fiscalyear_lock_date = vals.get('fiscalyear_lock_date') and\
+            fields.Date.from_string(vals['fiscalyear_lock_date'])
+        tax_lock_date = vals.get('tax_lock_date') and\
+            fields.Date.from_string(vals['tax_lock_date'])
```

Comment:
those lines cut are weird, and doesn't even help to stay in the 80-characters limit... I'd just  put everything in the same line